unreal/src/lib.rs

```diff
@@ -0,0 +1,201 @@
+pub enum Unreal4ParseError {
+    /// Expected UnrealEngine4 crash (zlib compressed)
+    #[fail(display = "unknown bytes format")]
+    UnknownBytesFormat,
+    /// Value out of bounds
+    #[fail(display = "out of bounds")]
+    OutOfBounds,
+    /// Invalid compressed data
+    #[fail(display = "bad compression")]
+    BadCompression(io::Error),
+}
+
+/// Unreal Engine 4 crash file
+#[derive(Debug)]
+pub struct Unreal4Crash {
+    bytes: Bytes,
+    files: Vec<CrashFileMeta>,
+}
+
+impl Unreal4Crash {
+    /// Creates an instance of `Unreal4Crash` from the original, compressed bytes
+    pub fn from_bytes(bytes: &[u8]) -> Result<Unreal4Crash, Unreal4ParseError> {
+        if bytes.is_empty() {
+            return Err(Unreal4ParseError::UnknownBytesFormat);
+        }
+
+        let mut zlib_decoder = zlib::Decoder::new(bytes);
+
+        let mut decompressed = Vec::new();
+        zlib_decoder
+            .read_to_end(&mut decompressed)
+            .map_err(Unreal4ParseError::BadCompression)?;
+
+        let decompressed = Bytes::from(decompressed);
+
+        let file_meta = get_files_from_bytes(&decompressed)?;
+
+        Ok(Unreal4Crash {
+            bytes: decompressed,
+            files: file_meta,
+        })
+    }
+
+    /// Files within the UE4 crash dump
+    pub fn files(&self) -> impl Iterator<Item = &CrashFileMeta> {
+        self.files.iter()
+    }
+
+    /// Count of files within the UE4 crash dump
+    pub fn file_count(&self) -> usize {
+        self.files.len() as usize
+    }
+
+    /// Get a `CrashFileMeta` by its index
+    pub fn file_by_index(&self, index: usize) -> Option<&CrashFileMeta> {
+        self.files().find(|f| f.index == index)
+    }
+
+    /// Get the contents of a file by its index
+    pub fn file_contents_by_index(&self, index: usize) -> Result<Option<&[u8]>, Unreal4ParseError> {
+        match self.file_by_index(index) {
+            Some(f) => Ok(Some(self.get_file_content(f)?)),
+            None => Ok(None),
+        }
+    }
+
+    /// Get the Minidump file bytes
+    pub fn get_minidump_bytes(&self) -> Result<Option<&[u8]>, Unreal4ParseError> {
+        let minidump = match self
+            .files()
+            // https://github.com/EpicGames/UnrealEngine/blob/5e997dc7b5a4efb7f1be22fa8c4875c9c0034394/Engine/Source/Runtime/Core/Private/GenericPlatform/GenericPlatformCrashContext.cpp#L60
+            .find(|f| f.file_name == "UE4Minidump.dmp")
+        {
+            Some(m) => m,
+            None => return Ok(None),
+        };
+
+        Ok(Some(self.get_file_content(minidump)?))
+    }
+
+    /// Get file content
+    pub fn get_file_content(&self, file_meta: &CrashFileMeta) -> Result<&[u8], Unreal4ParseError> {
+        let start = file_meta.offset;
+        let end = file_meta.offset + file_meta.len;
+
+        if self.bytes.len() < start || self.bytes.len() < end {
+            return Err(Unreal4ParseError::OutOfBounds);
+        }
+
+        Ok(&self.bytes[start..end])
```

Comment:
You can probably use `self.bytes.get(start..end)` with an `ok_or` instead of the manual bounds check. Also instead of `+` i would use `.saturating_add` or `.wrapping_add` to avoid the overflow panic.